### ISSUE TITLE
Two simple identical expressions result in different values on Linux32

### ISSUE THREAD

**Issue by daviditen** [chapel-lang/chapel#5598]
### Summary of Problem
On Linux32 platforms, the following code returns false

```
const sec1 = t.hour*3600 + t.minute*60 + t.second + t.microsecond / 1000000.0;
const sec2 = t.hour*3600 + t.minute*60 + t.second + t.microsecond / 1000000.0;
return sec1 == sec2;
```

### Steps to Reproduce
On a Linux32 platform:
`start_test test/expressions/diten/sameExprDiffValue.chpl`

**Associated Future Test(s):**
test/expressions/diten/sameExprDiffValue.future

### Configuration Information

- Output of `chpl --version`:
chpl Version 1.14.0.b686fd3

- Output of `$CHPL_HOME/util/printchplenv --anonymize`:
```
CHPL_TARGET_PLATFORM: linux32
CHPL_TARGET_COMPILER: gnu
CHPL_TARGET_ARCH: native
CHPL_LOCALE_MODEL: flat
CHPL_COMM: none
CHPL_TASKS: qthreads
CHPL_LAUNCHER: none
CHPL_TIMERS: generic
CHPL_UNWIND: none
CHPL_MEM: jemalloc
CHPL_MAKE: make
CHPL_ATOMICS: intrinsics
CHPL_GMP: gmp
CHPL_HWLOC: hwloc
CHPL_REGEXP: re2
CHPL_WIDE_POINTERS: struct
CHPL_AUX_FILESYS: none
```

- Back-end compiler and version, e.g. `gcc --version` or `clang --version`:
gcc (Ubuntu 4.9.1-16ubuntu6) 4.9.1

**Comment by e-kayrakli** [comment]
I was wondering if we have a new reason to believe this is a Chapel/denormalize issue. I thought it has been concluded that such comparisons were unsafe. I don't remember the details of the discussion we had back then, but I could find this:

https://chapel.atlassian.net/browse/CHAPEL-244

**Comment by daviditen** [comment]
I think the difference from CHAPEL-244 is that it was seeing differences across different platforms, which is somewhat to be expected.  In this case we're doing the same operations twice in a single compile and the operations should have a well defined order.  If the values and order of operations are the same between the two it should get the same result, right?

**Comment by ronawho** [comment]
I've think seen cases where something like:

```chpl
const sec1 = t.hour*3600 + t.minute*60 + t.second + t.microsecond / 1000000.0;
return sec1 == (t.hour*3600 + t.minute*60 + t.second + t.microsecond / 1000000.0);
```

will fail because the rhs computation is still in 80 bit (or however many bits) floating point register, while sec1 was truncated back to a 64 bit float. Wonder if copy-prop and/or denorm are causing something like this to happen. 

That said, I agree that if sec1 and sec2 are both being truncated into 64-bit floats, they _should_ be equal

**Comment by e-kayrakli** [comment]
I took a look at the issue a bit:

@ronawho is right. Denormalize creates something very similar to what he suggests. The underlying reason is because division primitive has side effects (possible divide-by-zero) and denormalize pass is very conservative when dealing with essential primitives.

For instance, changing that division to multiplication creates drastically different code (eliminates all temps) and lhs and rhs of `==` becomes identical. I haven't tested on linux32 but I assume the future would pass, if that division was multiplication.

**Comment by e-kayrakli** [comment]
I take the last comment back.. It doesn't get rid of everything and it would still fail even with multiplication. Although division is an issue, there is still something that denormalize is not able to work out...

**Comment by daviditen** [comment]
@tmacd9 points out that it also returns `false` when the entire operation is done as a single statement:

```
record T {
  var hour, minute, second, microsecond: int;
}

proc isEqualToSelf(t: T) {
  return ((t.hour*3600 + t.minute*60 + t.second + t.microsecond / 1000000.0) == (t.hour*3600 + t.minute*60 + t.second + t.microsecond / 1000000.0));
}

var a = new T(1,2,3,4);

writeln(isEqualToSelf(a)); 
```

**Comment by e-kayrakli** [comment]
I am interested in finding a solution to the issue, but I am afraid it is hard for me to do that for couple of weeks. Is there any urgency to solve this?

**Comment by daviditen** [comment]
I'm not blocked by it, so I IMO it's not urgent.

**Comment by daviditen** [comment]
This was fixed by PR #5980.
